syncmap.go

```diff
@@ -0,0 +1,171 @@
+package dapper
+
+import (
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/dogmatiq/iago/indent"
+	"github.com/dogmatiq/iago/must"
+)
+
+func mapFilter(
+	w io.Writer,
+	v Value,
+	f func(w io.Writer, v Value) error,
+) (err error) {
+	defer must.Recover(&err)
+
+	if v.IsAmbiguousType() {
+		must.WriteString(w, v.TypeName())
+	}
+	if v.Value.IsZero() {
+		must.WriteString(w, "{}")
+		return
+	}
+
+	i := syncMapItems{}
+	if v.Value.Addr().MethodByName("Range").Call(
+		[]reflect.Value{
+			reflect.ValueOf(i.populate(v, f)),
+		},
+	); i.Err != nil {
+		return i.Err
+	}
+
+	if len(i.Items) == 0 {
+		must.WriteString(w, "{}")
+		return
+	}
+
+	must.WriteString(w, "{\n")
+
+	i.print(
+		// TO-DO(danilvpetrov): inject the indenter into the filter, rather than
+		// hardcoding the indentation. Probably a 'Formatter' interface should
+		// be created that would look something like this:
+		//
+		// type Formatter interface {
+		//		Format(w io.Writer, v Value) error
+		// 		Indenter() io.Writer
+		// }
+		//
+		// Then, we could pass it as a third argument to all filters instead of
+		// f func(w io.Writer, v Value) error.
+		indent.NewIndenter(w, []byte("    ")),
+		f,
+	)
+
+	must.WriteString(w, "}")
+
+	return
+}
+
+type syncMapItem struct {
+	KeyWidth    int
+	KeyString   string
+	ValueString string
+}
+
+type syncMapItems struct {
+	Alignment int
+	Items     []syncMapItem
+	Err       error
+
+	alignToLastLine bool
+}
+
+func (m *syncMapItems) populate(
+	parent Value,
+	format func(w io.Writer, v Value) error,
+) func(interface{}, interface{}) bool {
```

Comment:
I think this can be refactored to avoid the need to store the error in the struct, let's look at it together when you're in the office.